### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by pbissonho**
## Status
READY

## Breaking Changes
YES

### Removed the need to implement the Token interface
Currently to use a custom token it is necessary to implement the Token interface, but the interface does not have any public method, which ends up creating difficulties. Now it is only necessary to define the generic <T>.
The Token class still continues to avoid another break change, but it’s no longer necessary to implement.
In the interface description itself it is defined as "Generic", so why not really make it generic?
```dart
/// {@template token}
/// Generic Token Interface
/// {@endtemplate}
abstract class Token {}
```

### Added the removeToken method
It is now possible to use removeToken () when it is necessary to update the AuthenticationStatus
to AuthenticationStatus.unauthenticated, instead of using setToken (null). 
That way, _tokenStorage.write (null) will no longer be called, but _tokenStorage.delete () end is possible to create a custom Token and TokenStorage in a simple way and make sure the storage has not received a null token.

### Added a property like authenticationStatus to return existing token

```dart
  Stream<T> get currentToken async* {
    yield _token;
    yield* _tokenController.stream;
  }
```
### Removed all statics attributes

This way the instances do not share the internal states. 

#### Implemented the Sink interface

As each instance has its own stream controllers, it is now possible to end with the "close" method.
Implemented the Sink interface to provide the closing method.In this way, it is also possible to set the token with the "add" method and users will be advised to close this by the "close_sinks" rule.

### OAuth2 constructor
To use OAuth2Token you will need to use the Fresh.oAuth2() constructor.
The default constructor now requires the tokenHeader.

### Internal improvements 
Internal changes were made to improve the code, facilitate maintenance and ensure that the FreshLink and FreshDio provide the same internal behavior.

## Todos
- [x] Tests
- [x] Documentation

## Objective
Improve lib usage.
Better the code base.
Facilitate the creation and maintenance of tests.
Improve documentation comments.

#### Perhaps it would be interesting to add an example of using FreshLink <T> to guarantee the expected behavior.

### CODE REVIEW

**Comment by felangel** [comment]
Hey @pbissonho 👋
Thanks for taking the time to open this PR! I’ve been super busy these last few days but will have a look this weekend 😊

**Comment by pbissonho** [comment]
> Hey @pbissonho 
> Thanks for taking the time to open this PR! I’ve been super busy these last few days but will have a look this weekend

Hello @felangel, take your time.
I saw that you worked hard in the last few days to release the Bloc v5.0.

File: packages/fresh/lib/src/fresh.dart [hunk 1]
```diff
@@ -89,3 +91,122 @@ class InMemoryTokenStorage<T extends Token> implements TokenStorage<T> {
...
+/// An interface that must be implemented to create an interceptor
+/// that transparently updates / caches tokens.
+abstract class FreshBase<T> {
+  /// {@template removeToken}
+  /// Sets the internal [token] to the provided [token]
+  /// and updates the `AuthenticationStatus` accordingly.
+  ///
+  /// If the provided token is null, the `AuthenticationStatus` will be updated
+  /// to `unauthenticated` and
+  /// and the token will be removed from storage, otherwise it
```
Comment (felangel) [inline packages/fresh/lib/src/fresh.dart:104]:
```suggestion
  /// the token will be removed from storage, otherwise it
```

File: packages/fresh/lib/src/fresh.dart [hunk 2]
```diff
@@ -89,3 +91,122 @@ class InMemoryTokenStorage<T extends Token> implements TokenStorage<T> {
...
+  FreshController({@required TokenStorage<T> tokenStorage})
+      : _tokenStorage = tokenStorage {
+    _tokenStorage.read().then(updateStatus);
+  }
+
+  final StreamController<AuthenticationStatus> _controller =
+      StreamController<AuthenticationStatus>.broadcast()
+        ..add(AuthenticationStatus.initial);
+
+  final StreamController<T> _tokenController = StreamController<T>.broadcast();
```
Comment (felangel) [inline packages/fresh/lib/src/fresh.dart:150]: we should provide a `close` in order to close these internal `StreamControllers`

File: packages/fresh_dio/.vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (felangel) [inline packages/fresh_dio/.vscode/launch.json:1]: can we revert this file?

File: packages/fresh_dio/lib/src/fresh.dart [hunk 4]
```diff
@@ -23,81 +25,89 @@ typedef RefreshToken<T> = Future<T> Function(T token, Dio httpClient);
...
-    _controller.add(authenticationStatus);
-    _token = token;
+  Future<void> removeToken() async {
+    await _freshController.removeToken();
   }
 
   @override
   Future<dynamic> onRequest(RequestOptions options) async {
     final token = await _getToken();
+    final data = _tokenHeader(token);
```
Comment (felangel) [inline packages/fresh_dio/lib/src/fresh.dart:100]:
```suggestion
    final tokenHeader = _tokenHeader(token);
```

File: packages/fresh_dio/lib/src/fresh.dart [hunk 5]
```diff
@@ -23,81 +25,89 @@ typedef RefreshToken<T> = Future<T> Function(T token, Dio httpClient);
...
+
     if (token != null) {
-      (options.headers ?? <String, String>{}).addAll(_tokenHeader(token));
+      (options.headers ?? <String, String>{}).addAll(data);
```
Comment (felangel) [inline packages/fresh_dio/lib/src/fresh.dart:103]:
```suggestion
      (options.headers ?? <String, String>{}).addAll(tokenHeader);
```

File: packages/fresh_dio/pubspec.yaml [hunk 6]
```diff
@@ -12,8 +12,12 @@ environment:
 dependencies:
   dio: ^3.0.9
   meta: ^1.1.8
-  fresh: ^0.2.1
-
+#  fresh:
+#    git: 
+#      url: https://github.com/pbissonho/fresh.git
+#      path: packages/fresh/
```
Comment (felangel) [inline packages/fresh_dio/pubspec.yaml:18]:
```suggestion
```

File: packages/fresh_dio/pubspec.yaml [hunk 7]
```diff
@@ -12,8 +12,12 @@ environment:
...
+  fresh:
+    path: ../fresh/
```
Comment (felangel) [inline packages/fresh_dio/pubspec.yaml:20]:
```suggestion
    path: ../fresh
```

Comment (felangel) [inline packages/fresh_dio/pubspec.yaml:20]:
```suggestion
    path: ../fresh
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 8]
```diff
@@ -115,14 +136,14 @@ void main() {
     });
 
     group('onRequest', () {
-      final oauth2Token = OAuth2Token(accessToken: 'accessToken');
+      final ooAuth2Token = OAuth2Token(accessToken: 'accessToken');
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:139]:
```suggestion
      final oAuth2Token = OAuth2Token(accessToken: 'accessToken');
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 9]
```diff
@@ -115,14 +136,14 @@ void main() {
...
       test(
-          'appends token header when token is OAuth2Token '
+          'appends token header when token is OoAuth2Token '
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:141]:
```suggestion
          'appends token header when token is OAuth2Token '
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 10]
```diff
@@ -115,14 +136,14 @@ void main() {
...
           'and tokenHeader is not provided', () async {
         final options = RequestOptions();
-        when(tokenStorage.read()).thenAnswer((_) async => oauth2Token);
+        when(tokenStorage.read()).thenAnswer((_) async => ooAuth2Token);
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:144]:
```suggestion
        when(tokenStorage.read()).thenAnswer((_) async => oAuth2Token);
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 11]
```diff
@@ -137,12 +158,12 @@ void main() {
       });
 
       test(
-          'appends token header when token is not OAuth2Token '
+          'appends token header when token is OoAuth2Token '
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:161]:
```suggestion
          'appends token header when token is OAuth2Token '
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 12]
```diff
@@ -137,12 +158,12 @@ void main() {
...
           'and tokenHeader is provided', () async {
         final options = RequestOptions();
-        when(tokenStorage.read()).thenAnswer((_) async => oauth2Token);
+        when(tokenStorage.read()).thenAnswer((_) async => ooAuth2Token);
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:164]:
```suggestion
        when(tokenStorage.read()).thenAnswer((_) async => oAuth2Token);
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 13]
```diff
@@ -158,18 +179,39 @@ void main() {
       });
 
       test(
-          'throws unimplemented if token is not OAuth2Token '
-          'and tokenHeader is not provided', () async {
+          'appends the standart header when token use OoAuth2Token constructor'
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:182]:
```suggestion
          'appends the standart header when token use OAuth2Token constructor'
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 14]
```diff
@@ -158,18 +179,39 @@ void main() {
...
+          'but tokenHeader is not provided', () async {
         final options = RequestOptions();
-        when(tokenStorage.read()).thenAnswer((_) async => MockToken());
+        when(tokenStorage.read()).thenAnswer((_) async => ooAuth2Token);
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:185]:
```suggestion
        when(tokenStorage.read()).thenAnswer((_) async => oAuth2Token);
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 15]
```diff
@@ -158,18 +179,39 @@ void main() {
...
           tokenStorage: tokenStorage,
           refreshToken: emptyRefreshToken,
         );
+        final actual = await fresh.onRequest(options) as RequestOptions;
+        expect(
+          actual.headers,
+          {
+            'content-type': null,
+            'authorization':
+                '${ooAuth2Token.tokenType} ${ooAuth2Token.accessToken}',
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:197]:
```suggestion
                '${oAuth2Token.tokenType} ${oAuth2Token.accessToken}',
```

File: packages/fresh_dio/test/fresh_test.dart [hunk 16]
```diff
@@ -396,7 +441,7 @@ void main() {
     });
   });
 
-  group('OAuth2Token', () {
+  group('OoAuth2Token', () {
```
Comment (felangel) [inline packages/fresh_dio/test/fresh_test.dart:444]:
```suggestion
  group('OAuth2Token', () {
```

File: packages/fresh_dio/lib/src/fresh.dart [hunk 17]
```diff
@@ -23,81 +25,89 @@ typedef RefreshToken<T> = Future<T> Function(T token, Dio httpClient);
...
+  ///
+  /// ```dart
+  /// dio.interceptors.add(
+  ///   Fresh.oAuth2Token(
+  ///     tokenStorage: InMemoryTokenStorage<OAuth2Token>(),
+  ///     refreshToken: (token, client) async {...},
+  ///   ),
+  /// );
+  /// ```
+  static Fresh<OAuth2Token> oAuth2Token({
```
Comment (felangel) [inline packages/fresh_dio/lib/src/fresh.dart:59]:
```suggestion
  static Fresh<OAuth2Token> oAuth2({
```

File: packages/fresh_graphql/lib/src/fresh_link.dart [hunk 18]
```diff
@@ -26,121 +29,115 @@ typedef RefreshToken<T> = Future<T> Function(T, Client);
...
+  ///     // Perform refresh and return new token
+  ///   },
+  /// );
+  /// final graphQLClient = GraphQLClient(
+  ///   cache: InMemoryCache(),
+  ///   link: Link.from([freshLink, HttpLink(uri: 'https://my.graphql.api')]),
+  /// );
+  /// ```
+  /// {@endtemplate}
+  static FreshLink<OAuth2Token> oAuth2Token({
```
Comment (felangel) [inline packages/fresh_graphql/lib/src/fresh_link.dart:66]:
```suggestion
  static FreshLink<OAuth2Token> oAuth2({
```

File: packages/fresh_graphql/lib/src/fresh_link.dart [hunk 19]
```diff
@@ -26,121 +29,115 @@ typedef RefreshToken<T> = Future<T> Function(T, Client);
...
-  /// Returns a `Stream<AuthenticationState>` which is updated internally based
-  /// on if a valid token exists in [TokenStorage].
-  Stream<AuthenticationStatus> get authenticationStatus async* {
-    yield _authenticationStatus;
-    yield* _controller.stream;
+  ///{@template fresh_link}
+  ///A GraphQL Link which handles manages an authentication token automatically.
+  ///
+  /// ```dart
+  /// final freshLink = FreshLink.oAuth2Token(
```
Comment (felangel) [inline packages/fresh_graphql/lib/src/fresh_link.dart:54]:
```suggestion
  /// final freshLink = FreshLink.oAuth2(
```

File: packages/fresh_dio/lib/src/fresh.dart [hunk 20]
```diff
@@ -23,81 +25,89 @@ typedef RefreshToken<T> = Future<T> Function(T token, Dio httpClient);
...
-  static final StreamController _controller =
-      StreamController<AuthenticationStatus>.broadcast()
-        ..add(AuthenticationStatus.initial);
+  /// {@macro freshInterceptor}.
+  /// A constructor that returns a Fresh interceptor that uses the
+  /// `OAuth2Token` token, the standard token class.
+  ///
+  /// ```dart
+  /// dio.interceptors.add(
+  ///   Fresh.oAuth2Token(
```
Comment (felangel) [inline packages/fresh_dio/lib/src/fresh.dart:53]:
```suggestion
  ///   Fresh.oAuth2(
```

**Comment by pbissonho** [comment]
> Overall the changes look great but had a few comments. Let me know what you think and thanks again for taking the time to make these changes -- I really appreciate it

Nice comments. I agree with all the suggested changes and I will resolve them.

File: packages/fresh/lib/src/fresh.dart (hunk 1 above)
Comment (pbissonho) [reply packages/fresh/lib/src/fresh.dart]: Fixed.

File: packages/fresh_dio/.vscode/launch.json (hunk 3 above)
Comment (pbissonho) [reply packages/fresh_dio/.vscode/launch.json]: File removed.

File: packages/fresh_dio/lib/src/fresh.dart (hunk 4 above)
Comment (pbissonho) [reply packages/fresh_dio/lib/src/fresh.dart]: Renamed.

File: packages/fresh_dio/pubspec.yaml (hunk 6 above)
Comment (pbissonho) [reply packages/fresh_dio/pubspec.yaml]: Fixed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 8 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 10 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 12 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 14 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 15 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 16 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Renamed.

File: packages/fresh_dio/lib/src/fresh.dart (hunk 17 above)
Comment (pbissonho) [reply packages/fresh_dio/lib/src/fresh.dart]: Renamed.

File: packages/fresh_graphql/lib/src/fresh_link.dart (hunk 19 above)
Comment (pbissonho) [reply packages/fresh_graphql/lib/src/fresh_link.dart]: Renamed.

File: packages/fresh_dio/lib/src/fresh.dart (hunk 5 above)
Comment (pbissonho) [reply packages/fresh_dio/lib/src/fresh.dart]: Done.

File: packages/fresh_dio/pubspec.yaml (hunk 7 above)
Comment (pbissonho) [reply packages/fresh_dio/pubspec.yaml]: Done.

File: packages/fresh_dio/test/fresh_test.dart (hunk 9 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Fixed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 11 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Fixed.

File: packages/fresh_dio/test/fresh_test.dart (hunk 13 above)
Comment (pbissonho) [reply packages/fresh_dio/test/fresh_test.dart]: Fixed.

File: packages/fresh_graphql/lib/src/fresh_link.dart (hunk 18 above)
Comment (pbissonho) [reply packages/fresh_graphql/lib/src/fresh_link.dart]: Renamed.

File: packages/fresh_dio/lib/src/fresh.dart (hunk 20 above)
Comment (pbissonho) [reply packages/fresh_dio/lib/src/fresh.dart]: Fixed.

File: packages/fresh/lib/src/fresh.dart (hunk 2 above)
Comment (pbissonho) [reply packages/fresh/lib/src/fresh.dart:162]: I really forgot to provide this method.
I implemented the Sink interface to provide the closing method.In this way, it is also possible to add the token with the "add" method and  users will be advised to close this by the close_sinks rule.

Let me know what you think.

File: packages/fresh_graphql/example/main.dart [hunk 21]
```diff
@@ -1,40 +0,0 @@
-import 'dart:io';
```
Comment (felangel) [inline packages/fresh_graphql/example/main.dart:1]: why did we delete the example?

File: packages/fresh/analysis_options.yaml [hunk 22]
```diff
@@ -9,3 +9,5 @@ linter:
 analyzer:
   exclude:
     - test/.test_coverage.dart
+  errors:
+    deprecated_member_use: ignore
```
Comment (felangel) [inline packages/fresh/analysis_options.yaml:13]: why is this needed?

File: packages/fresh_graphql/example/main.dart (hunk 21 above)
Comment (pbissonho) [reply packages/fresh_graphql/example/main.dart]: I accidentally deleted this file. Recovered file.

File: packages/fresh/analysis_options.yaml (hunk 22 above)
Comment (pbissonho) [reply packages/fresh/analysis_options.yaml:13]: Since I defined the Token class as depreciated, the analyze dart is causing the checks to fail. Add to make all checks pass.
As the package will have a considerable break change, should we already remove the Token class?

Comment (felangel) [reply packages/fresh/analysis_options.yaml]: yup will just publish 0.3.0 👍
